ftw/solr/upgrades/20200401111512_update_modified_index/upgrade.py

```diff
@@ -0,0 +1,163 @@
+from time import clock
+from zope.component import getMultiAdapter
+from zope.component import queryUtility
+import logging
+import transaction
+from ftw.solr.browser.maintenance import checkpoint_iterator
+from ftw.solr.browser.maintenance import timer
+
+
+def _pre_datetime_format_fix_to_iso8601(value, multivalued=False):
+    if isinstance(value, DateTime):
+        v = value.toZone('UTC')
+        value = u'%04d-%02d-%02dT%02d:%02d:%06.3fZ' % (
+            v.year(), v.month(), v.day(), v.hour(), v.minute(), v.second()
+        )
+    elif isinstance(value, datetime):
+        if value.tzinfo is not None:
+            # Convert to timezone naive in UTC
+            value = (value - value.utcoffset()).replace(tzinfo=None)
+
+        value = u'%04d-%02d-%02dT%02d:%02d:%02d.%03dZ' % (
+            value.year, value.month, value.day, value.hour, value.minute,
+            value.second, value.microsecond / 1000
+        )
+    elif isinstance(value, date):
+        value = u'%04d-%02d-%02dT%02d:%02d:%02d.%03dZ' % (
+            value.year, value.month, value.day, 0, 0, 0, 0
+        )
+    else:
+        value = None
+    return value
+
+
+def _solr_date(date):
+    """Solr date representation. Fractional seconds are stripped if 0."""
+    value = to_iso8601(date)
+    if value.endswith('.000Z'):
+        value = value[:-5] + 'Z'
+    return value
+
+
+def _pre_datetime_format_fix_solr_date(date):
+    """Solr date representation. Fractional seconds are stripped if 0."""
+    value = _pre_datetime_format_fix_to_iso8601(date)
+    if value.endswith('.000Z'):
+        value = value[:-5] + 'Z'
+    return value
+
+
+commit_interval = 100
+
+
+class UpdateModifiedIndex(UpgradeStep):
+    """Update modified index.
+    """
+
+    def __call__(self):
+        self.logger = logging.getLogger('ftw.upgrade')
+
+        if not self.is_enabled():
+            self.log("Skipping: Solr is not enabled.")
+            return
+        self.sync()
+
+    def is_enabled(self):
+        registry = queryUtility(IRegistry)
+        settings = registry.forInterface(ISolrSettings)
+        return settings.enabled
+
+    def get_objects_that_need_syncing_of_modified(self):
+        """Find objects that are not in sync only due to change in DateTime
+        rounding"""
+
+        catalog = getToolByName(self.portal, 'portal_catalog')
+        items = catalog.unrestrictedSearchResults()
+        catalog_modified = set(
+            [(item.UID, _solr_date(item.modified)) for item in items])
+        pre_datetime_format_fix_catalog_modified = set(
+            [(item.UID, _pre_datetime_format_fix_solr_date(item.modified)) for item in items])
+
+        conn = self.manager.connection
+        resp = conn.search({
+            u'query': u'*:*',
+            u'limit': 10000000,
+            u'params': {u'fl': ['UID', 'modified']},
+        })
+        solr_modified = set(
+            [(doc['UID'], doc.get('modified', u'2000-01-01T00:00:00.000Z'))
```

Comment:
not sure why `'2000-01-01T00:00:00.000Z'` is the default. but it is the same where you copied if from so probably best keep it.

Reply:
yes I think it should behave as much as possible the same as `diff`